doc/md/knowledge-center/mysql/generated_columns.md

```diff
@@ -0,0 +1,218 @@
+slug: /knowledge/mysql/generated-columns
+---
+
+**MySQL** is a popular open-source relational database. **Generated columns** are a feature of MySQL that allows you to define tables with columns whose value is a function of the value stored in other columns; without requiring complex expressions in SELECT, INSERT or UPDATE queries.
+
+## What are Generated Columns?
+
+Generated columns are columns that contain values calculated by expressions which can be dependent on other columns; in a similar manner to formulas in a spreadsheet. There are two types of generated columns in MySQL: _Stored_ and _Virtual_.
+
+### Stored Generated Columns
+
+Stored generated columns are stored and evaluated when a row is inserted or updated. As a result, stored generated columns use disk space in addition to CPU cycles during the execution of `INSERT` and `UPDATE` statements.
+
+### Virtual Generated Columns
+
+Virtual generated columns are not stored, and only evaluated when a row is read _(after BEFORE [triggers](https://dev.mysql.com/doc/refman/5.7/en/trigger-syntax.html))_. As a result, virtual generated columns take no storage at the cost of CPU cycles for `SELECT` statements.
+
+### Limitations of Generated Columns
+
+Generated column expressions must be deterministic which means that — given the same input — an expression must always produce the same output. As a result, generated columns can not be used with stored variables, functions, procedures, and subqueries; which could cause the output to be non-deterministic. Following this constraint, generated columns can not be used to generate random values. On the other hand, a generated column may reference any non-generated column _regardless_ of its position within the table row and any other generated column within the same table row, as long as those columns are declared before the generated column.
+
+## When to use Generated Columns?
+
+Generated columns should be used whenever you want to create a column with a value that can be directly determined from the values of other columns in the same row. In simpler words, for data that is dependent on other data. This saves the developer from complex application code that is prone to errors on `SELECT`, `INSERT` and `UPDATE` statements. It also ensures that data which must be consistent, stays consistent.
+
+**MySQL Syntax for a Generated Column**
+
+```sql
+column_name data_type [GENERATED ALWAYS] AS (expr) [VIRTUAL | STORED] 
+    [NOT NULL | NULL] [UNIQUE [KEY]] [[PRIMARY] KEY] [COMMENT 'string']
+```
+
+### Using Stored Generated Columns
+
+Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.
```

Comment:
That is for virtual columns and is already mentioned in the virtual column section. A virtual column can only be used with secondary indexes. Please see https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html#:~:text=A%20stored%20column%20does%20require%20storage%20space%20and%20can%20be%20indexed.

Reply:
When referring to indexes, it's usually secondary indexes (keys). The primary index is the primary key (index). Did you mean this? 

 > Stored generated columns should also be used when you want to index the column

When reading this, I got the impression that I should use "STORED" (and not "VIRTUAL") in case I want to index the column.

Reply:
Yes. If you want to use a secondary index (which allows duplicates), you can use stored or virtual columns. However, if you want to use a primary index (no duplicates), then you must use a stored column.

Reply:
OK, can we change "index" to "primary key" in this sentence? As I mentioned above, it's common in the "industry" to call secondary indexes/keys just indexes, and primary indexes as primary keys.

Reply:
An index isn't necessarily a primary key, but a primary key is an index.
- https://dev.mysql.com/doc/refman/8.0/en/create-index.html
- https://dev.mysql.com/doc/refman/8.0/en/mysql-indexes.html

I was told to prioritize correctness over simplicity with regards to these types of sentences, so by that objective the sentence should remain the same.

Reply:
From reading this sentence, I'll expect the following command to not pass:

<img width="649" alt="image" src="https://user-images.githubusercontent.com/7413593/177418240-aa5afb77-47ac-48c8-b9da-fbfa80418f3a.png">

As you can see, I created a **virtual** column and indexed it.

Reply:
The article later states:

> If you use the _InnoDB Storage Engine_, secondary indexes can be defined on virtual columns.

The focus of this comment is on the use-case for stored generated columns:

> Stored generated columns should also be used when you want to index the column

This doesn't imply that the following command will fail, but only that a virtual column is only valid with a secondary index, while a stored generated column can be used with any index.

Feel free to suggest a change to make this distinction clearer. However, specifying this difference in the stored generated column section would be wrong. For example, "Stored generated columns should also be used when you want to index the column _(except for secondary indexes)_" is an incorrect statement. You can use any index with a stored generated column, but only secondary indexes on the virtual column _(with the InnoDB Storage Engine)_.